### ISSUE TITLE
Rule camelcase cannot detect between upper-camelcase and lower-camelcase when lint `property`.

### ISSUE THREAD

**Issue by tinymins** [eslint/eslint#10473]
**What rule do you want to change?**

camelcase

**Does this change cause the rule to produce more or fewer warnings?**

More warnings.

**How will the change be implemented? (New option, new default behavior, etc.)?**

New options.

```js
{
    propertiesStyle: "upper" | "lower" | "all",
}
```

**Please provide some example code that this change will affect:**

This is bad code example.
```js
const camel = {};
camel.UpperCamelcase = true;
```

This is good code example.
```js
const camel = {};
camel.lowerCamelcase = true;
```

**What does the rule currently do for this code?**

All will be ok. The rule currently do not detect what type of camelcase is using for `property`.

**What will the rule do after it's changed?**

There will be an option to configure upper-camelcase or lower-camelcase for `property`.

**What's more**

I'm willing to fork and make a PR if my idea is acceptable. 😃 

https://github.com/tinymins/eslint/tree/properties-style

**Comment by platinumazure** [comment]
Can `id-match` be used in this case? (Legitimately not sure and wanted to ask.)

**Comment by tinymins** [comment]
@platinumazure I'm sorry that I've reply to wrong thread just now. Actually `id-match` has problems when limit lower camelcase for object property name.

**Comment by tinymins** [comment]
![image](https://user-images.githubusercontent.com/1808990/41349999-3270e98a-6f44-11e8-9d74-578215c2c5de.png)
@platinumazure Here's the screenshot. Use `id-match` in current version cannot limit lower-camelcase in property. Rule `id-match` also need to be updated to satisfy this situation.

**Comment by tinymins** [comment]
Or should I turn to concentrate on fix `id-match` bugs?

**Comment by platinumazure** [comment]
No, I was only asking if it could work for your use case.

Let's see what the rest of the team thinks.

**Comment by tinymins** [comment]
Thanks, I've tried that.

If `id-match` rule can separate object property pattern from commonly pattern.

**Comment by not-an-aardvark** [comment]
My understanding is that capital letters at the start of a name are conventionally used for constructor functions, whereas lowercase letters are conventionally used for almost everything else. If this option were implemented, do you think you would use lower-camelcase letters for constructor names? To me, it seems like the problem is that ESLint is allowing upper-camelcase names for non-constructor-function values, not that it's allowing upper-camelcase names in general. (Unfortunately, the former problem probably wouldn't be fixable due to dynamic typing.)

**Comment by tinymins** [comment]
@not-an-aardvark Yes, just like what you said, upper-camelcase means it is a constructor.
But constructor usually does not appears in an object.
```js
// We usually use constructor like this:
import ClassA from 'class-a';
const a = new ClassA();

// Hardly ever seen this:
// obj is defined somewhere
const a = obj.ClassA();
```

So we want to lint object properties name in lower-camelcase.

Thanks a lot for reply.

**Comment by tinymins** [comment]
It seems that I should stop thinking about modify rule `camelcase`, to be concentrated on fixing `id-match` bugs?

**Comment by tinymins** [comment]
Hi everyone, I'v forked and made some improvements for rule `id-match`. Several bugs has been fixed in the days just past. I think maybe we should remove underscores ignore codes in `camelcase`.

Thanks every for keeping focus on this issue, here's my new PR: https://github.com/eslint/eslint/pull/10477

**Comment by mcclowes** [comment]
Hey @tinymins, will you have a chance to split your PR (https://github.com/eslint/eslint/pull/10477) soon? Would be good to have this resolved! :)

**Comment by tinymins** [comment]
Sorry that I got busy on my work and had a cold last week. I'm working on it now. @mcclowes

**Comment by tinymins** [comment]
New PR created, I don't know if there will be a conflict. If there is, I'll do a git rebase job. @mcclowes

**Comment by tinymins** [comment]
https://github.com/eslint/eslint/pull/10554
